ynput/OpenPype
# AYON: Use folder path as unique identifier

## Changelog Description
Use folder path instead of asset name as unique identifier, with OpenPype compatibility.

## Additional info
Goal of this PR is to support duplicated folder names in a project for AYON mode. How it works? All places that have asset name as "unique identifier" and not as "name" value are using folder path now. Some places do/did use the value for other purposes, all of those places must use different approach.

### What it affects
IMPORTANT FOR DEVELOPERS
- `get_current_asset_name` returns path in AYON mode
- `os.environ["AVALON_ASSET"]` and `legacy_io.Session["AVALON_ASSET"]` contains path in AYON mode
- during publishing `context.data["asset"]` and `instance.data["asset"]` contains path in AYON mode
- all pipeline functions that expect asset name as unique identifier (e.g. `get_template_data_with_names`) expect path instead of name in AYON mode
- creators (except legacy creators) should use `"folderPath"` instead of `"asset"`
- `openpype.client` functions expecting asset name are can handle folder path in AYON mode > the functions still can find by name
    - `get_assets`
    - `get_archived_assets`
    - `get_asset_by_name`
    - `get_last_version_by_subset_name`

### TODOs
- folder path will start with slash in future version of AYON server, so that must be changed acordingly
- editorial publishing was not modified at all (I'm not capable of doing that)
    - [hiero](https://github.com/ynput/OpenPype/pull/5829)
    - flame
    - resolve
    - traypublisher
- revisit Harmony `openpype/hosts/harmony/plugins/load/load_template.py` line 55 > There is passed asset name but I don't know what for
- revisit `ExtractUnrealSkeletalMeshFbx` plugin in Maya which is using asset name and not sure for what
- 3dsmax has some mismatched data in `CollectRender` plugin (different PR to resolve [In progress])
- houdini `CreateHDA` has one occation where it cannot work with current implementation
- unreal has at multiple places mismatch with OpenPype asset name and Unreal asset name
- host integrations without publisher (using legacy create tool) are not converted
    - blender
    - harmony

## Testing notes:
### Folder path have to start with slash in on AYON server
- backend PR https://github.com/ynput/ayon-backend/pull/56

### Allow to create duplicated names on WebUI
- frontend branch [enhancement/allow_duplicate_folders](https://github.com/ynput/ayon-frontend/tree/enhancement/allow_duplicate_folders)

### Testing
1. Have AYON project which allows duplicated names.
    - To allow this you have to modify database on running AYON docker stack
    a) Go to your ayon-docker repository.
    b) Run `./manage.ps1 dbshell` or `make dbshell`
    c) Execute `DROP INDEX project_{project name}.folder_unique_name;` where `{project name}` is replaced by name of your project
2. Create duplicated names in the project hierarchy
3. Run DCCs on a folder with duplicated name

- Try to use all creators and publish them, and validate the output ended up at correct places.
    - Make sure the template of `{asset}` or `{folder[name]}` is NOT filled with path.
- Validate all tools > visual aspects.
- Report any issues (check if the issue was reported first).

- open scene with created instances before this PR -> what happens in Publisher

### Hosts
- [x] AfterEffects @kalisp 
- [x] Blender @simonebarbieri - needs revisit after https://github.com/ynput/OpenPype/pull/5910
- [ ] <strike>Flame @jakubjezek001 </strike> Won't be converted at this point.
- [x] Fusion @jakubjezek001 
- [ ] <strike>Harmony @kalisp </strike> Will be dealt with in own addon.
- [x] Hiero @jakubjezek001 https://github.com/ynput/OpenPype/pull/5829
- [x] Houdini @MustafaJafar 
- [x] 3dsMax @moonyuet 
- [x] Maya @antirotor 
- [x] Nuke @mkolar
- [x] Photoshop @kalisp
- [x] Resolve @jakubjezek001  https://github.com/ynput/OpenPype/pull/5892
- [x] Substance Painter @LiborBatek 
- [x] TrayPublisher @iLLiCiTiT
- [x] TrayPublisher - editorial @jakubjezek001 https://github.com/ynput/OpenPype/pull/5873
- [x] TVPaint @iLLiCiTiT 
- [x] Unreal @simonebarbieri
